File: ts/src/includes.ts
import { realpathSync } from "node:fs";
import { dirname, extname, resolve } from "node:path";

import type { Reference } from "./edit.ts";
import { EntryconfError } from "./errors.ts";
import type { Loader } from "./loader.ts";
import { decodeUtf8, formatForExtension, parseDocument } from "./parse.ts";
import { escapePointerToken } from "./pointer.ts";
import { emptyObject, setKey, type Value } from "./tree.ts";

export const INCLUDE_PREFIX = "@file:";

/**
 * The include resolver's position: the file whose value is being walked and
 * its directory (paths are relative to the referencing file), the pointer
 * within that file (`src`) and within the effective tree (`eff`), the real
 * paths of the files being resolved (cycle detection, innermost last), and the
 * `@file:` references traversed so far (SPEC §10.3 "chain", absolute paths
 * until `open` keys them).
 */
export interface Walk {
  doc: string;
  dir: string;
  src: string;
  eff: string;
  files: string[];
  chain: Reference[];
}

function step(w: Walk, token: string): Walk {
  const t = `/${escapePointerToken(token)}`;
  return { ...w, src: w.src + t, eff: w.eff + t };
}

/** Whether an authored string is an `@file:` reference. */
export function isInclude(s: string): boolean {
  return s.startsWith(INCLUDE_PREFIX);
}

/** The cleaned absolute path an `@file:` string names, relative to `dir`. */
export function includeTarget(s: string, dir: string): string {
  return resolve(dir, s.slice(INCLUDE_PREFIX.length));
}

/**
 * Replace every `@file:<path>` string with the parsed tree of its target and
 * unescape leading `@@` (SPEC §5).
 */
export function resolveIncludes(l: Loader, value: Value, w: Walk): Value {
  if (typeof value === "string") return resolveString(l, value, w);
  if (Array.isArray(value)) {
    return value.map((item, i) => resolveIncludes(l, item, step(w, String(i))));
  }
  if (typeof value === "object" && value !== null) {
    const out = emptyObject();
    for (const [key, item] of Object.entries(value)) {
      setKey(out, key, resolveIncludes(l, item, step(w, key)));
    }
    return out;
  }
  return value;
}

function resolveString(l: Loader, value: string, w: Walk): Value {
  if (value.startsWith("@@")) {
    // A doubled leading "@" is a literal "@"; the result is never an include.
    return value.slice(1);
  }
  if (isInclude(value)) {
    return includeFile(l, value.slice(INCLUDE_PREFIX.length), w);
  }
  if (value.startsWith("@")) {
    throw new EntryconfError(
      "E_SUBSTITUTION",
      `${JSON.stringify(value)} is not a valid directive; write "@@" for a literal leading "@"`,
    );
  }
  return value;
}

function includeFile(l: Loader, target: string, w: Walk): Value {
  if (target === "") {
    throw new EntryconfError("E_INCLUDE", "empty @file: path");
  }
  const path = resolve(w.dir, target);
  const format = formatForExtension(extname(path));
  if (format === null) {
    throw new EntryconfError(
      "E_INCLUDE",
      `${path}: unsupported extension (expected .json, .yaml, .yml, or .toml)`,
    );
  }

  // Cycles are detected on real paths, so a file reached through two names
  // is still one file; everything else uses the path as referenced.
  let real: string;
  try {
    real = l.src.isFile(path) ? realpathSync(path) : path;
  } catch {
    throw new EntryconfError("E_INCLUDE", `${path}: missing or unreadable`);
  }
  const seen = w.files.indexOf(real);
  if (seen !== -1) {
    throw new EntryconfError(
      "E_INCLUDE_CYCLE",
      `include cycle: ${[...w.files.slice(seen), real].join(" -> ")}`,
    );
  }

  let bytes: Uint8Array;
  try {
    bytes = l.src.readFile(path);
  } catch {
    throw new EntryconfError("E_INCLUDE", `${path}: missing or unreadable`);
  }
  // A missing or unreadable target is E_INCLUDE, but content that is not valid
  // UTF-8 is a parse fault like any other (SPEC §2, §5).
  const tree = parseDocument(decodeUtf8(bytes, path), format, path);
  const chain: Reference[] = [...w.chain, { document: w.doc, pointer: w.src }];
  if (l.rec) {
    l.rec.record(path, bytes, format, tree);
    l.rec.graft(path, { effective: w.eff, chain });
  }
  return resolveIncludes(l, tree, {
    doc: path,
    dir: dirname(path),
    src: "",
    eff: w.eff,
    files: [...w.files, real],
    chain,
  });
}

File: python/src/entryconf/_includes.py
"""``@file:`` include grafting (SPEC §5)."""

from __future__ import annotations

import os
from dataclasses import dataclass, replace
from pathlib import Path
from typing import Any

from ._errors import E_INCLUDE, E_INCLUDE_CYCLE, E_PARSE, E_SUBSTITUTION, EntryconfError
from ._parsers import SUFFIXES, parse_document
from ._pointer import escape_token
from ._source import clean_join, format_of

_PREFIX = "@file:"


def is_include(value: str) -> bool:
    return value.startswith(_PREFIX)


def include_target(value: str, base_dir: Path) -> Path:
    """The cleaned path an ``@file:`` string names, relative to its file's directory."""
    return clean_join(base_dir, value[len(_PREFIX) :])


@dataclass
class Walk:
    """Where the resolver is: the file being walked and its directory, the
    pointer within that file (``src``) and within the effective tree
    (``eff``), the realpath chain for cycle detection, and the ``@file:``
    references traversed so far (SPEC §10.3 "chain")."""

    doc: Path
    base_dir: Path
    src: str
    eff: str
    files: list[Path]
    chain: list[dict[str, str]]

    def step(self, token: str) -> "Walk":
        escaped = "/" + escape_token(token)
        return replace(self, src=self.src + escaped, eff=self.eff + escaped)


class Resolver:
    def __init__(self, src: Any, rec: Any = None) -> None:
        self.src = src
        self.rec = rec

    def read_document(self, path: Path, missing_code: str) -> Any:
        """Read and parse one document; read failures use ``missing_code``."""
        try:
            data = self.src.read_file(path)
            text = data.decode("utf-8")
        except OSError as exc:
            raise EntryconfError(missing_code, f"{path}: {exc}") from exc
        except UnicodeDecodeError as exc:
            raise EntryconfError(E_PARSE, f"{path}: {exc}") from exc
        tree = parse_document(path, text)
        if self.rec is not None:
            self.rec.record(path, data, format_of(path), tree)
        return tree

    def _graft(self, value: str, w: Walk) -> Any:
        target = include_target(value, w.base_dir)
        # SPEC §5: the extension is matched case-sensitively; `.JSON` is not `.json`.
        if target.suffix not in SUFFIXES:
            raise EntryconfError(
                E_INCLUDE, f"{target}: unsupported extension {target.suffix!r}"
            )
        real = Path(os.path.realpath(target))
        if real in w.files:
            cycle = [str(p) for p in w.files[w.files.index(real) :]] + [str(real)]
            raise EntryconfError(E_INCLUDE_CYCLE, " -> ".join(cycle))
        tree = self.read_document(target, E_INCLUDE)
        chain = w.chain + [{"document": str(w.doc), "pointer": w.src}]
        if self.rec is not None:
            self.rec.graft(target, w.eff, chain)
        return self.resolve(
            tree,
            Walk(doc=target, base_dir=target.parent, src="", eff=w.eff, files=w.files + [real], chain=chain),
        )

    def resolve(self, node: Any, w: Walk) -> Any:
        """Replace every ``@file:`` string value below ``node`` with its tree."""
        if isinstance(node, str):
            if node.startswith("@@"):
                # A leading `@@` becomes a literal `@`; the result is inert.
                return "@" + node[2:]
            if node.startswith(_PREFIX):
                return self._graft(node, w)
            if node.startswith("@"):
                raise EntryconfError(
                    E_SUBSTITUTION, f"{node!r}: reserved `@` directive (write `@@` for a literal `@`)"
                )
            return node
        if isinstance(node, dict):
            return {key: self.resolve(value, w.step(key)) for key, value in node.items()}
        if isinstance(node, list):
            return [self.resolve(item, w.step(str(i))) for i, item in enumerate(node)]
        return node
